rclcpp_components/include/rclcpp_components/component_manager_isolated.hpp

```diff
@@ -81,14 +78,36 @@ class ComponentManagerIsolated : public rclcpp_components::ComponentManager
   {
     auto executor_wrapper = dedicated_executor_wrappers_.find(node_id);
     if (executor_wrapper != dedicated_executor_wrappers_.end()) {
-      executor_wrapper->second.promise.set_value();
-      executor_wrapper->second.executor->cancel();
-      executor_wrapper->second.thread.join();
+      cancel_executor(executor_wrapper->second);
       dedicated_executor_wrappers_.erase(executor_wrapper);
     }
   }
 
 private:
+  /// Stops a spinning executor avoiding race conditions.
+  /**
+   * @param executor_wrapper executor to stop and its associated thread
+   */
+  void cancel_executor(DedicatedExecutorWrapper & executor_wrapper)
+  {
+    // We can't immediately call the cancel() API on an executor because if it is not
+    // already spinning, this operation will have no effect.
+    // We rely on the assumption that this class creates executors and then immediately makes them
+    // spin in a separate thread, i.e. the time gap between when the executor is created and when
+    // it starts to spin is small (although it's not negligible).
+
+    while (!executor_wrapper.executor->is_spinning()) {
+      // This is an arbitrarily small delay to avoid busy looping
+      rclcpp::sleep_for(std::chrono::milliseconds(1));
+    }
```

Comment:
I would love to have something better than this, but it also doesn't look too bad.
The only problem I see if the spinning thread crashes.

Reply:
Yes, I agree that it's not elegant, but at the same time it's also very simple.
What about adding a timeout here?

There isn't currently any try/catch around spin, so if that thread crashes, the whole app will crash